### ISSUE TITLE
redundant_self_in_closure false positive

### ISSUE THREAD

**Issue by la-pieuvre** [realm/SwiftLint#5010]
### New Issue Checklist

- [x ] Updated SwiftLint to the latest version
- [x ] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

in the typical case where a guard let is used :

```swift
guard let lastTimerCallTime else {return}
self.lastTimerCallTime = Date()
```
affecting a new value to the object property "self.lastTimerCallTime" trigger a false positive and remove the self creating an error since local lastTimerCallTime is immutable

### Environment

* SwiftLint version 0.52.2
* Installation method used Homebrew
* Paste your configuration file:

```yml
opt_in_rules:
 - redundant_self_in_closure
```

* Which Xcode version are you using ? 14.3

```swift
guard let lastTimerCallTime else {return}
self.lastTimerCallTime = Date()
```

**Comment by SimplyDanny** [comment]
This is a known shortcoming of this rule that must generally be accepted, unfortunately. The reason is that normal SwiftLint rules only operate on the Swift AST without knowing anything about types.

The line `self.lastTimerCallTime = lastTimerCallTime` would be the only special case that is allowed. The rule doesn't trigger on the `self`. Besides your example there are a bunch of other conceivable cases that cannot easily and especially fast be detected, however.

**Comment by la-pieuvre** [comment]
Thank you for your fast answer. 
What about detecting the value assignment combined with the guard let to prevent the self to be remove?
